R/tostring.R

```diff
@@ -1,3 +1,244 @@
+#'   \item{`spacewidth`}{(`numeric(1)`)\cr the width of the space character in the currently active font.}
+#'   \item{`ismonospace`}{(`logical(1)`)\cr whether the specified font is monospaced.}
+#' }
+#'
+#' `open_font_dev` opens a pdf device with the specified font
+#' only if there is not one currently open with the same font.
+#' If a new device is opened, it caches `spacewidth` and
+#' `ismonospace` for use in `nchar_ttype`).
+#'
+#' `close_font_dev` closes any open font state device
+#' and clears the cached values.
+#'
+#' @return
+#' - `open_font_dev` returns a logical value indicating whether a *new* pdf device was opened.
+#' - `close_font_dev` returns `NULL`.
+#'
+#' In both cases the value is returned invisibly.
+#'
+#' @examples
+#' open_font_dev(font_spec("Times"))
+#' nchar_ttype("Hiya there", font_spec("Times"))
+#' close_font_dev()
+#'
+#' @export
+open_font_dev <- function(fontspec, silent = FALSE) {
+  if (is.null(fontspec)) {
+    return(invisible(FALSE))
+  } else if (font_dev_state$open) {
+    if (identical(font_dev_state$fontspec, fontspec)) {
+      if (!silent && dev.cur() != font_dev_state$dev_num) {
+        warning(
+          "formatters is switching to the font state graphics device ",
+          "to perform string width calculations. You may need to switch ",
+          "to your currently open graphics device, depending on whether ",
+          "the font device is closed and what other devices you have open."
+        )
+        dev.set(font_dev_state$dev_num)
+      }
+      return(invisible(FALSE))
+    } else {
+      close_font_dev()
+    }
+  } else if (FALSE && !font_dev_state$open) { ## remove 'FALSE &&' to get debugging info to helplocate places which aren't receiving/using the state properly # nolint
+    # start nocov
+    ## dump the call stack any time we have cache misses
+    ## and have to open a completely new font state device
+    scalls <- sys.calls()
+    msg <- sapply(
+      scalls[2:length(scalls)],
+      function(sci) {
+        toret <- deparse(sci[[1]], nlines = 3)
+        if (substr(toret[1], 1, 8) == "function") {
+          toret <- "anon function"
+        }
+        toret
+      }
+    )
+    cat("\n***** START font dev debugging dump *****\n")
+    cat(paste(msg, collapse = " -> "), "\n")
+    print(fontspec)
+  } #end nocov
+  tmppdf <- tempfile(fileext = ".pdf")
+  pdf(tmppdf)
+  grid.newpage()
+  gp <- gpar_from_fspec(fontspec)
+  pushViewport(plotViewport(gp = gp))
+  spcwidth <- cwidth_inches_unsafe(" ")
+  ## XXX this assumes M or W is the widest "reasonable char"
+  ## should be true for any reasonable font but not guaranteed!!!
+  mwidth <- cwidth_inches_unsafe("M")
+  wwidth <- cwidth_inches_unsafe("W")
+  assign("open", TRUE, envir = font_dev_state)
+  assign("fontspec", fontspec, envir = font_dev_state)
+  assign("spacewidth", spcwidth, envir = font_dev_state)
+  assign("ismonospace", spcwidth == cwidth_inches_unsafe("W"),
```

Comment:
why not `wwidth`?

Reply:
thats a logical. its checking (and caching the answer) whether " " and "W" are the same width (if they are its assumed monospace, though I Suppose technically one could artificially construct a font where "W" is the width of " " but other characters are wider or shorter... I dont' think we need to protect ourself against that eventuality though

Reply:
Oh, nm I misunderstood. 

Yeah, we could use `wwidth` except that I am removing those cause none of that stuff is used (except for by the max ratio calculation which itself is not used anywhere. :)